dnprogs/dndel/dndel.1
.TH DNDEL 1 "October 2 1998" "DECnet utilities"

.SH NAME
dndel \- Delete a file on a VMS system
.SH SYNOPSIS
.B dndel
[options] file-name
.br
Options:
.br
[\-ivVh]
.SH DESCRIPTION
.PP
.B dndel
deletes files held on VMS systems. The filename parameter may specify a single
file or a wildcard (remember to use VMS wildcards).

If you specify 
.B \-i
or
.B \-v
then dndel will open two connections to the VMS machine (much as VMS itself 
does). Otherwise it will attempt to use a single connection. One side-effect
of this is that the deletion process will fail at the first file that cannot
be deleted.
.SH OPTIONS
.TP
.I "\-i"
Interactive. Prompt before deleting a file.
.TP
.I "\-v"
Verbose. Print the names of files that have been deleted
.TP
.I "\-T connect timeout"
Specifies the maximum amount of time the command will wait to establish a connection
with the remote node. a 0 here will cause it to wait forever. The default is 60 seconds
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of the tools package that dndel comes from.

.SH EXAMPLES

  dndel \-i 'myvax::oldfile;*'

.br
  dndel 'tramp"christine pjc123"::interactive'
.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dncopy "(1), " dntask "(1) ," dnsubmit "(1), " dnprint "(1)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/dndel/dndel.cc
/******************************************************************************
    (c) 1998-2008      Christine Caulfield          christine.caulfield@googlemail.com

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 ******************************************************************************
 */
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <sys/types.h>
#include <string.h>
#include <errno.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include <unistd.h>
#include <regex.h>

#include "connection.h"
#include "protocol.h"
#include "logging.h"


static void usage(void)
{
    printf("\nUSAGE: dndel [OPTIONS] 'node\"user password\"::filespec'\n\n");
    printf("NOTE: The VMS filename really should be in single quotes to\n");
    printf("      protect it from the shell\n");

    printf("\nOptions:\n");
    printf("  -i           interactive - prompt before deleting\n");
    printf("  -v           verbose - display files that have been deleted\n");
    printf("  -T <secs>    connect timeout (default 60)\n");
    printf("  -? -h        display this help message\n");
    printf("  -V           show version number\n");

    printf("\nExamples:\n\n");
    printf(" dndel -l 'myvax::*.*;'\n");
    printf(" dndel 'cluster\"christine thecats\"::disk$users:[cats.pics]otto.jpg;'\n");
    printf("\n");
}


// Check for an extra STATUS message following a NAME
static bool check_status(dap_connection &conn, char *name)
{
  dap_message *m = dap_message::read_message(conn, false);
  if (m)
  {
      if (m->get_type() == dap_message::STATUS)
      {
	  dap_status_message *sm = (dap_status_message *)m;
	  fprintf(stderr, "Error deleting %s: %s\n", name, sm->get_message());
	  delete m;
	  return false;
      }

      // The file must exist if we get an ATTRIB message.
      if (m->get_type() == dap_message::ATTRIB)
      {
	  delete m;
	  return true;
      }

      fprintf(stderr, "Got unexpected message: %s\n", m->type_name());
      delete m;
  }
  return true;
}


// Send an ACCESS DIRECTORY request. If we are running non-interactively then
// we actually delete the files here instead.
static	bool dap_directory_lookup(dap_connection &c, char *dirname,
				  bool interactive)
{
    dap_access_message acc;
    if (interactive)
	acc.set_accfunc(dap_access_message::DIRECTORY);
    else
	acc.set_accfunc(dap_access_message::ERASE);
    acc.set_accopt(1);
    acc.set_filespec(dirname);
    acc.set_display(dap_access_message::DISPLAY_MAIN_MASK);
    return acc.write(c);
}

// Delete one file
static bool dap_delete_file(dap_connection &c, char *name)
{
    dap_access_message acc;
    acc.set_accfunc(dap_access_message::ERASE);
    acc.set_accopt(1);
    acc.set_filespec(name);
    acc.set_display(0);
    return acc.write(c);
}

// Get a response from the delete request
static bool dap_get_delete_ack(dap_connection &c)
{
    dap_message *m = dap_message::read_message(c, true);
    if (m)
    {
	switch (m->get_type())
	{
	case dap_message::ACCOMP:
	case dap_message::ACK:
	    return true;

	case dap_message::STATUS:
	    {
		dap_status_message *sm = (dap_status_message *)m;
		fprintf(stderr, "Error deleting file: %s\n", sm->get_message());
		return false;
	    }
	}
    }
    return false;
}

// Send CONTRAN/SKIP message. We need this if a file in the list is locked.
static bool dap_send_skip(dap_connection &conn)
{
    dap_contran_message cm;
    cm.set_confunc(dap_contran_message::SKIP);
    return cm.write(conn);
}

// Get the file information. Actually just the name
static int dap_get_dir_entry(dap_connection &c, char *name)
{
    static char volume[256];
    static char dir[256];
    static bool got_name = false;

    dap_message *m;
    while ( ((m = dap_message::read_message(c, true))) )
    {
	switch (m->get_type())
	{
	case dap_message::NAME:
	    {
		dap_name_message *nm = (dap_name_message *)m;
		switch (nm->get_nametype())
		{
		case dap_name_message::VOLUME:
		    strcpy(volume, nm->get_namespec());
		    break;

		case dap_name_message::DIRECTORY:
		    strcpy(dir, nm->get_namespec());
		    break;

		case dap_name_message::FILENAME:
		    sprintf(name, "%s%s%s", volume, dir, nm->get_namespec());
		    got_name = true;
		    delete m;
		    if (check_status(c, name))
		        return dap_message::NAME;
		    else
		        return -1;

		case dap_name_message::FILESPEC:
		    strcpy(name, nm->get_namespec());
		    got_name = true;
		    delete m;
		    if (check_status(c, name))
		        return dap_message::NAME;
		    else
		        return -1;
		}

	    }
	    break;

	case dap_message::ACCOMP:
	    delete m;
	    return -1;

	case dap_message::ACK:
	    delete m;
	    return dap_message::ACK;

	case dap_message::STATUS:
	    {
		dap_status_message *sm = (dap_status_message *)m;
		if (sm->get_code() == 0x4030) // Locked
		{
		    dap_send_skip(c);
		    break;
		}
		fprintf(stderr, "Error deleting %s: %s\n",
			got_name?name:"file",
			sm->get_message());
		delete m;
		return -1;
	    }

	default:
	    delete m;
	    return m->get_type();
	}
	delete m;
    }
    return -1;
}

// Start here...
int main(int argc, char *argv[])
{
    int	    opt,retval;
    char    name[256];
    bool    interactive = false;
    int     verbose = 0;
    int     two_links = 0;
    int     connect_timeout = 60;

    if (argc < 2)
    {
	usage();
	exit(0);
    }

/* Get command-line options */
    opterr = 0;
    optind = 0;
    while ((opt=getopt(argc,argv,"?hvViT:")) != EOF)
    {
	switch(opt)
	{
	case 'h':
	case '?':
	    usage();
	    exit(1);

	case 'i':
	    interactive = true;
	    two_links++;
	    break;

	case 'T':
	    connect_timeout = atoi(optarg);
	    break;

	case 'v':
	    verbose++;
	    two_links++;
	    break;

	case 'V':
	    printf("\ndndel from dnprogs version %s\n\n", VERSION);
	    exit(1);
	    break;
	}
    }
    init_logging("dndel", 'e', false);

    dap_connection dir_conn(verbose);
    dap_connection del_conn(verbose);

    dir_conn.set_connect_timeout(connect_timeout);
    del_conn.set_connect_timeout(connect_timeout);

    /* We open one link to get the file names and (if interactive or verbose)
       another to do the actual deletion.
       This is not a waste of resources - it's what VMS does !
     */
    if (!dir_conn.connect(argv[optind], dap_connection::FAL_OBJECT, name))
    {
	fprintf(stderr, "%s\n", dir_conn.get_error());
	return -1;
    }
    // Exchange config messages
    if (!dir_conn.exchange_config())
    {
	fprintf(stderr, "Error in config: %s\n", dir_conn.get_error());
	return -1;
    }

    /* Open the second link to actually delete the file(s) */
    if (two_links)
    {
	if (!del_conn.connect(argv[optind], dap_connection::FAL_OBJECT, name))
	{
	    fprintf(stderr, "%s\n", del_conn.get_error());
	    return -1;
	}
	// Exchange config messages
	if (!del_conn.exchange_config())
	{
	    fprintf(stderr, "Error in config: %s\n", del_conn.get_error());
	    return -1;
	}
    }

    /* If non-interactive then the next command just deletes the files */
    dap_directory_lookup(dir_conn, name, interactive || (verbose>0) );

    /* Loop through the files we find */
    while ((retval=dap_get_dir_entry(dir_conn, name)) > 0)
    {
	if (retval == dap_message::NAME)
	{
	    if (interactive)
	    {
		char response[255];

		printf("Delete %s ? ", name);
		fgets(response, sizeof(response), stdin);
		if (tolower(response[0]) == 'y')
		{
		    dap_delete_file(del_conn, name);
		    if (dap_get_delete_ack(del_conn))
			if (verbose)
			    printf("Deleted %s\n", name);
		}
	    }
	    else if (verbose)
	    {
		dap_delete_file(del_conn, name);
		if (dap_get_delete_ack(del_conn))
		    printf("Deleted %s\n", name);
	    }
	}
    }

    dir_conn.close();
    if (two_links) del_conn.close();
    return 0;
}

